Make PrivCount Tor Events Consistent

teor2345 opened the pull request:
Consistently count bytes, cells, streams, circuits, and connections.
Consistently exclude directory (including BEGINDIR), origin, and (for streams) non-exit circuits.

Stubs out unused fields: they will be removed in #244.

teor2345 committed:
Set is_dir flag based on the linked connection's type

Closes #190, but further fixes to come in #191.

teor2345 committed:
Accurately identify client connections by their lack of peer authentication

This means that we bail out earlier if asked to extend to a client.

Follow-up to Tor Trac 21407.
Fixes Tor bug 21406; bugfix on 0.2.4.23.

Cherry-picked from 46fe353f25 on Tor 0.3.1.0-alpha-dev.
Modified to remove references to ed25519 link authentication introduced
in Tor 0.3.0.
Fixes #199 in PrivCount Tor.

teor2345 committed:
Stop sending events for BEGINDIR data, streams, circuits, and connections

Tor uses a separate connection, circuit, and stream for one-hop directory
connections, so this avoids counting most directory data.

But tor also gets and posts some directory documents over 3-hop connections:
cells and bytes for these connections is still included in the counters.

Part of #230.

teor2345 committed:
Make PrivCount consistently count and log events

A rewrite and refactor that simplifies the PrivCount code, while also
resolving several issues:
* Consistently count bytes, cells, streams, circuits, and connections
  (If a stream, circuit, or connection is ignored, the equivalent bytes
  and cells are also ignored.)
* Exclude BEGINDIR and DirPort connections
  Fixes #230
  Fixes #190
  Still needs cleanup to remove excess fields in the events, see #244.
* Identify DNS resolves correctly, but include them as long as
  PRIVCOUNT_DNS_RESOLVE_IS_OVERHEAD is not defined as 1.
  Fixes #238
  Still needs cleanup to remove excess fields in the events, see #244.
* Identify relays and clients correctly
  Fixes #199
  Still needs cleanup to remove excess fields in the events, see #244.
* Stop confusing previous and next circuit remote addresses
  Fixes #243

robgjansen commented on src/or/control.c:
```diff
@@ -5794,88 +5794,459 @@ control_event_bandwidth_used(uint32_t n_read, uint32_t n_written)
+/* Should PrivCount count streams and bytes from this connection or circuit?
+ */
+int
+privcount_is_counted_for_bytes(const edge_connection_t* exitconn,
```
I found the `is_counted_for` a bit awkward. How about something like `privcount_should_count_bytes`?

teor2345 replied:
Fixed in aea93e83e by using "privcount_data_is" rather than "privcount_is" for all applicable functions. This particular function is now named:
privcount_data_is_used_for_byte_counters

teor2345 replied:
I made the names even clearer in 812a65b.

robgjansen commented on src/or/control.c:
```diff
@@ -5794,88 +5794,459 @@ control_event_bandwidth_used(uint32_t n_read, uint32_t n_written)
+/* Should PrivCount ignore cells, circuits, and connections from this circuit?
+ */
+int
+privcount_is_counted_for_cells(const connection_t *conn,
```
I suggest `privcount_should_count_cells`.

teor2345 replied:
I fixed this, see my comment above in https://github.com/privcount/tor/pull/4#discussion_r113611850

robgjansen commented on src/or/control.c:
```diff
@@ -5794,88 +5794,459 @@ control_event_bandwidth_used(uint32_t n_read, uint32_t n_written)
+/* Should PrivCount ignore DNS requests from this connection or circuit?
+ */
+int
+privcount_is_counted_for_dns(const edge_connection_t* exitconn,
```
I suggest `privcount_should_count_dns`.

teor2345 replied:
I fixed this, see my comment above in https://github.com/privcount/tor/pull/4#discussion_r113629202

robgjansen commented on src/or/control.c:
```diff
@@ -5794,88 +5794,459 @@ control_event_bandwidth_used(uint32_t n_read, uint32_t n_written)
+  }
+
+  if (!EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_CIRCUIT_ENDED) &&
+      !EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_CONNECTION_ENDED)) {
```
I wonder if the privcount side would change the "rules" here, then do we need to patch Tor again later to update this if statement? Perhaps that OK.

teor2345 replied:
The rules are set by the C code: we only increment the counters for events that are active. (We also only increment stream byte counters for exit connections. They are reported as 0 for non-exit connections, I opened #248 to fix this when we do Onion Service statistics.)

I agree this is confusing, and there is a bug in it: the byte counts should be done for the transfer, stream, and circuit events, and the cell counts should be done for the circuit event. I fixed this bug, and a few other errors in these checks, in 812a65bbf.

robgjansen commented on src/or/control.c:
```diff
@@ -5794,88 +5794,459 @@ control_event_bandwidth_used(uint32_t n_read, uint32_t n_written)
+  }
+
+  if (!EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_STREAM_BYTES_TRANSFERRED) &&
+      !EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_STREAM_ENDED)) {
```
I assume PrivCount would complain about the config if asked to count stream bytes but not streams, or otherwise would automatically enable both event types.

teor2345 replied:
PrivCount uses a union of the events needed for all counters, and any other events specified in the code. (The tests specify a custom set of events to make sure that all events work.)

robgjansen commented on src/or/control.c:
```diff
@@ -5794,88 +5794,459 @@ control_event_bandwidth_used(uint32_t n_read, uint32_t n_written)
+
+  /* DNS connections are never directory connections */
+  int is_dir = privcount_is_dir(TO_CONN(exitconn), TO_CIRCUIT(orcirc));
+  tor_assert_nonfatal(!is_dir);
```
Shouldn't we return 0 here if is_dir is true, meaning that we should not count DNS on this connection because it is a directory connection?

teor2345 replied:
The overhead function returns 1 when is_dir is true, and so we will return 0 here.

But if we are calling this function on a directory connection, something is very wrong with our code. This is why I added a non-fatal assert (we'll need to check our tor logs to pick these up).

I kept the same logic in 812a65b when I simplified this function.

robgjansen commented on src/or/control.c:
```diff
@@ -5794,88 +5794,459 @@ control_event_bandwidth_used(uint32_t n_read, uint32_t n_written)
+privcount_is_consensus_relay(const channel_t* chan)
+{
+  if (!chan) {
+    /* Assume that missing connections *aren't* a known relay */
```
Grammar nit pick - should be either `Assume that missing connections *aren't* known relays` or `Assume that a missing connection *isn't* a known relay`.

teor2345 replied:
This comment was moved and rewritten in aea93e8.

robgjansen commented on src/or/control.c:
```diff
@@ -5794,88 +5794,459 @@ control_event_bandwidth_used(uint32_t n_read, uint32_t n_written)
+privcount_is_unauthenticated_client(const channel_t *chan)
+{
+  if (!chan) {
+    /* Assume that missing connections *aren't* a client */
```
Grammar nit pick - should be either `Assume that missing connections *aren't* clients` or `Assume that a missing connection *isn't* a client`

teor2345 replied:
This comment was moved and rewritten in aea93e8.

robgjansen commented on src/or/control.c:
```diff
@@ -5794,88 +5794,459 @@ control_event_bandwidth_used(uint32_t n_read, uint32_t n_written)
+
+/* Perform a += b without overflow, and without returning a value */
+void
+privcount_sum(uint64_t *total, uint64_t increment) {
```
I'm not sure why this is safer/cleaner than just calling `privcount_add_saturating` directly from wherever `privcount_sum` is currently called. I realize it is a bit shorter because you don't have to do the assignment in the caller. Maybe it's just programmer style, but I usually avoid passing pointers and then operating on them when unnecessary (like when passing ints).

teor2345 replied:
You're right, passing a pointer does need an extra check. I had added that in aea93e8. But we really don't need to use pointers at all. And they inhibit compiler optimisations. Fixed in a8da3d2a9.

robgjansen commented on src/or/control.c:
```diff
@@ -5794,88 +5794,459 @@ control_event_bandwidth_used(uint32_t n_read, uint32_t n_written)
-        }
-        /* now we know its an or_circuit_t */
-        orcirc = TO_OR_CIRCUIT(circ);
+    /* There's no time here. We should fix that if we ever use this event */
```
What does "There's no time here." mean? It isn't being included in the control event string? If so, it seems trivial to add it.

teor2345 replied:
I opened privcount/privcount#249 for this change.
Because it's an interface change, we should do it when we change the interface in privcount/privcount#244 and privcount/privcount#178.

robgjansen reviewed (commented):
Looks great! Modularizing all of the logic about what to count was a great idea, and makes the actual control event code easier to understand.

I think pretty much all of the comments are minor and you can consider them optional.

teor2345 committed:
Make sure that cell counts don't overflow

This cell-specific code fixes #195
The refactoring (in multiple commits) fixes #186

teor2345 committed:
Document NULL handling and improve NULL checks in privcount functions

teor2345 committed:
Rename functions so their purpose is clearer

privcount_is_counted_for_X -> data_is_used_for_X_counters
privcount_is_X -> privcount_data_is_X

teor2345 committed:
Rewrite the PrivCount counter and event check code

Fixes the following bugs:
* Increment byte and cell counters for the correct events
* Trigger events under the right conditions

Non-functional improvements:
* Make function names even clearer
* Remove unused arguments
* Simplify the call graph of the checking functions
* Refactor to remove duplicate code

teor2345 committed:
Use privcount_add_saturating directly, rather than passing pointers

teor2345 committed:
fixup! Accurately identify client connections by their lack of peer authentication

teor2345 commented:
Hi @robgjansen, can you please review the commits I have pushed since https://github.com/privcount/tor/pull/4#commits-pushed-bb0d164

Once you have reviewed them, I will squash the branch and merge it.
This branch contains all the bug fixes I expect in this release.
We have 2 field additions and then a bunch of mass mechanical changes, like spacing, licensing, and field removal.

robgjansen reviewed (approved):
No complaints here :)

teor2345 commented:
This was tested with privcount/privcount#254 and works fine. Merging.
